### PR TITLE
KEP for co-scheduling plugin based on scheduler framework

### PR DESCRIPTION
**Pull request by denkensk**
Add co-scheduling plugin for the new scheduler framework.
This plugin will be put in kubernetes-sigs/scheduler-plugins firstly.
Kep for #1

/sig scheduling
/assign @Huang-Wei

### CODE REVIEW

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+---
+title: Coscheduling plugin based on scheduler framework
```
Comment (Huang-Wei) [inline kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: To distinguish with other coscheduling implementation, can we reword it to "Lightweight coscheduling based on back-to-back queue sorting"? And also rename this file as well.

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+      - [UnReserve](#unreserve)
+  - [Alternatives considered](#alternatives-considered)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Testing Plan](#testing-plan)
+  - [Implementation History](#implementation-history)
+  - [References](#references)
+<!-- /toc -->
+
+## Motivation
+Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads (ML/DL) are managed in a different controller. To improve cluster utilization and operation efficiency, we'd like to treat Kubernetes as a unified management platform. But ML jobs are All-or-Nothing: they require all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks to be ready to begin to work. In the worst case, all jobs are pending leading to a deadlock. To solve this problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible.
```
Comment (Huang-Wei) [inline kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: ...are managed in a different controller -> ...are managed in a different system.

But ML jobs... -> Given that ML jobs...

The new scheduler framework makes the goal possible. -> This proposal tries to solve this problem by implementing a coscheduling plugin based on scheduler framework.

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+}
+ 
+type PodGroupInfo struct {
+    // InitialTimestamp stores the timestamp of the initialization time of PodGroup.
+    InitialTimestamp 	time.Time
+    UID           			types.UID
+    MinAvailable  		int
+    Name                                 String
+}
+```
```
Comment (Huang-Wei) [inline kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: Can you reformat the indents?

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+- If they are both regularPods (without particular PodGroup label), compare their `InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned ahead of the other.
+  
+- If one is regularPod and the other is pgPod, compare regularPod's `InitialAttemptTimestamp` with pgPod's `InitialTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
+  
+- If they are both pgPods:
+  - Compare their `InitialAttemptTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
+  - If their `InitialAttemptTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `InitialAttemptTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)
+
+#### Pre-Filter
```
Comment (Huang-Wei) [inline kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]:
```suggestion
#### PreFilter
```

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+#### Permit
+In `Permit` phase, we put the pod that doesn't meet `minAvailable` into the WaitingMap and reserve resources until `minAvailable` are met or timeout is triggered.
+1. Get the number of Running pods that belong to the same PodGroup
+2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
+3. If Running + WaitingPods + 1 >= `minAvailable`(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: the timeout is dynamic value depends on the size of the `PodGroup`.)
+
+#### UnReserve
+After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.
+
```
Comment (Huang-Wei) [inline kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: I think we should add a "Known Limitations" section here to list any known limitations we've discussed in the [k/k KEP](https://github.com/kubernetes/enhancements/pull/1463), for example: how the Pods get moved back to activeQ ([comment](https://github.com/kubernetes/enhancements/pull/1463#discussion_r393711437))

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md (hunk 1 above)
Comment (denkensk) [reply kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: Done

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md (hunk 2 above)
Comment (denkensk) [reply kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: Done

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md (hunk 4 above)
Comment (denkensk) [reply kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: Done

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md (hunk 5 above)
Comment (denkensk) [reply kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: Done

File: kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md (hunk 3 above)
Comment (denkensk) [reply kep/20200116-coscheduling-plugin-based-on-scheduler-framework.md]: Done

**Comment by Huang-Wei** [comment]
@denkensk LGTM.

@ahg-g @alculquicondor Do you want to take another look? (We're going to review and host scheduler-plugins proposal in this subrepo)
